src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs

```diff
@@ -53,16 +67,27 @@ public sealed partial class NamedPipeServerStream : PipeStream
                 maxNumberOfServerInstances = 255;
             }
 
-            Interop.Kernel32.SECURITY_ATTRIBUTES secAttrs = PipeStream.GetSecAttrs(inheritability);
-            SafePipeHandle handle = Interop.Kernel32.CreateNamedPipe(fullPipeName, openMode, pipeModes,
-                maxNumberOfServerInstances, outBufferSize, inBufferSize, 0, ref secAttrs);
+            Interop.Kernel32.SECURITY_ATTRIBUTES secAttrs = PipeStream.GetSecAttrs(inheritability, pipeSecurity, out object pinningHandle);
 
-            if (handle.IsInvalid)
+            try
             {
-                throw Win32Marshal.GetExceptionForLastWin32Error();
-            }
+                SafePipeHandle handle = Interop.Kernel32.CreateNamedPipe(fullPipeName, openMode, pipeModes,
+                    maxNumberOfServerInstances, outBufferSize, inBufferSize, 0, ref secAttrs);
 
-            InitializeHandle(handle, false, (options & PipeOptions.Asynchronous) != 0);
+                if (handle.IsInvalid)
+                {
+                    throw Win32Marshal.GetExceptionForLastWin32Error();
+                }
+
+                InitializeHandle(handle, false, (options & PipeOptions.Asynchronous) != 0);
+            }
+            finally
+            {
+                if (pinningHandle != null)
```

Comment:
If we pass the GCHandle by ref to GetSecAttrs this can be changed to if (pinningHandle.IsAllocated) pinningHandle.Free();